nntrainer/layers/mol_attention_layer.cpp

```diff
@@ -116,19 +202,159 @@ void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
   u_base.add_i(-0.5);
   Tensor u_neg = u_base;
 
-  Tensor integral_left, integral_right;
-  sigmoid.run_fn(u_pos.subtract(state).divide(beta.add(1e-8f)), integral_left);
-  sigmoid.run_fn(u_neg.subtract(state).divide(beta.add(1e-8f)), integral_right);
-  Tensor integral = integral_left.subtract(integral_right);
+  Tensor beta_eps = beta.add(1e-8f);
+
+  Tensor u_pos_m = u_pos.subtract(m);
+  u_pos_m.divide(beta_eps, u_pos_div);
+  sigmoid.run_fn(u_pos_div, prob_left);
 
-  Tensor integral_scaled = integral.multiply(alpha);
-  Tensor scores = integral_scaled.sum(3);
-  scores.reshape(TensorDim({scores.batch(), 1, 1, scores.height()}));
+  Tensor u_neg_m = u_neg.subtract(m);
+  u_neg_m.divide(beta_eps, u_neg_div);
+  sigmoid.run_fn(u_neg_div, prob_right);
+
+  prob_left.subtract(prob_right, prob);
+
+  Tensor prob_scaled = prob.multiply(alpha);
+  prob_scaled.sum(3, scores);
 
   scores.dotBatched(value, output);
 }
 
-void MoLAttentionLayer::calcDerivative(RunLayerContext &context) { /** NYI */
+void MoLAttentionLayer::calcDerivativeHelper(RunLayerContext &context,
+                                             Tensor &dstate) {
+  Tensor &query = context.getInput(wt_idx[AttentionParams::query]);
+  Tensor &value = context.getInput(wt_idx[AttentionParams::value]);
+
+  Tensor &derivative = context.getIncomingDerivative(SINGLE_INOUT_IDX);
+
+  Tensor &fc_proj_out = context.getTensor(wt_idx[AttentionParams::fc_proj_out]);
+  Tensor &dfc_proj_out =
+    context.getTensor(wt_idx[AttentionParams::fc_proj_out]);
+  Tensor &scores = context.getTensor(wt_idx[AttentionParams::scores]);
+  Tensor &prob = context.getTensor(wt_idx[AttentionParams::prob]);
+  Tensor &prob_left = context.getTensor(wt_idx[AttentionParams::prob_left]);
+  Tensor &prob_right = context.getTensor(wt_idx[AttentionParams::prob_right]);
+  Tensor &u_neg_div = context.getTensor(wt_idx[AttentionParams::u_neg_div]);
+  Tensor &u_pos_div = context.getTensor(wt_idx[AttentionParams::u_pos_div]);
+
+  const TensorDim &input_dim = query.getDim();
+  unsigned int batch = input_dim.batch();
+  auto mol_k = std::get<props::MoL_K>(mol_props).get();
+
+  Tensor kappa, beta, alpha;
+  kappa.copy_with_stride(
+    fc_proj_out.getSharedDataTensor({batch, 1, 1, mol_k}, 0, false));
+  beta.copy_with_stride(
+    fc_proj_out.getSharedDataTensor({batch, 1, 1, mol_k}, mol_k, false));
+  alpha.copy_with_stride(
+    fc_proj_out.getSharedDataTensor({batch, 1, 1, mol_k}, mol_k * 2, false));
+
+  Tensor dscores = Tensor(TensorDim({value.batch(), 1, 1, value.height()}));
+  dscores.dot_batched_deriv_wrt_1(value, derivative);
+  dscores.reshape(TensorDim({scores.batch(), 1, scores.width(), 1}));
+
+  Tensor dprob_scaled = Tensor(TensorDim({batch, 1, value.height(), mol_k}));
+  dprob_scaled.setZero();
+  dprob_scaled.add_i(dscores);
```

Comment:
Throwing some list for double checking
1. scores.batch() == batch ?
2. scores.width() == values.height()?

Reply:
1. Skipped the batch size check as batch is assumed to be same across all inputs.
2. scores.width() is manually set to values.height() :) - https://github.com/nnstreamer/nntrainer/blob/775d4418ad208bcd9ba2a76d6c67f8c685a1c97e/nntrainer/layers/mol_attention_layer.cpp#L108

But the dimensions checks were completely missed in finalizing. Added dimensionality checks for values and query.

Reply:
👍